// tls-client-node/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TLSClientError = void 0;
function getErrorMessage(value) {
    if (value instanceof Error) {
        return value.message;
    }
    if (typeof value === "string") {
        return value;
    }
    try {
        return JSON.stringify(value);
    }
    catch {
        return String(value);
    }
}
class TLSClientError extends Error {
    code;
    status;
    details;
    retriable;
    constructor(message, options = {}) {
        super(message, { cause: options.cause });
        this.name = "TLSClientError";
        this.code = options.code;
        this.status = options.status;
        this.details = options.details;
        this.retriable = options.retriable;
    }
    static fromUnknown(value, options = {}) {
        if (value instanceof TLSClientError) {
            return value;
        }
        return new TLSClientError(getErrorMessage(value), {
            ...options,
            cause: options.cause ?? value,
        });
    }
}
exports.TLSClientError = TLSClientError;
