// tls-client-node/dist/native.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ensureNativeBinding = ensureNativeBinding;
const binary_1 = require("./binary");
const errors_1 = require("./errors");
const bindingCache = new Map();
function parseNativeResponse(rawResponse, operation, binding) {
    if (!rawResponse) {
        throw new errors_1.TLSClientError(`tls-client native ${operation} returned no data.`, {
            code: "ERR_NATIVE_EMPTY_RESPONSE",
        });
    }
    let parsed;
    try {
        parsed = JSON.parse(rawResponse);
    }
    catch (error) {
        // Best-effort: extract and free the Go allocation even from malformed JSON
        // to prevent memory leaks when the library returns a corrupted response.
        const idMatch = /"id"\s*:\s*"([^"]+)"/.exec(rawResponse);
        if (idMatch?.[1]) {
            binding.freeMemory(idMatch[1]);
        }
        throw new errors_1.TLSClientError(`tls-client native ${operation} returned invalid JSON: ${rawResponse}`, {
            code: "ERR_NATIVE_INVALID_RESPONSE",
            cause: error,
        });
    }
    if (parsed.id) {
        binding.freeMemory(parsed.id);
    }
    return parsed;
}
function invokeNativeAsync(operation, fn, ...args) {
    return new Promise((resolve, reject) => {
        fn.async(...args, (error, result) => {
            if (error) {
                reject(new errors_1.TLSClientError(`tls-client native ${operation} failed: ${error.message}`, {
                    code: "ERR_NATIVE_CALL_FAILED",
                    cause: error,
                }));
                return;
            }
            if (result === undefined) {
                reject(new errors_1.TLSClientError(`tls-client native ${operation} returned no data.`, {
                    code: "ERR_NATIVE_EMPTY_RESPONSE",
                }));
                return;
            }
            resolve(result);
        });
    });
}
async function loadNativeBinding(libraryPath) {
    const existing = bindingCache.get(libraryPath);
    if (existing) {
        return existing;
    }
    const koffiModule = (await import("koffi"));
    const load = koffiModule.load ?? koffiModule.default?.load;
    if (!load) {
        throw new errors_1.TLSClientError("Failed to load koffi for the Windows native backend.", {
            code: "ERR_NATIVE_BINDING_LOAD",
        });
    }
    const library = load(libraryPath);
    const rawBinding = {
        request: library.func("request", "string", ["string"]),
        getCookiesFromSession: library.func("getCookiesFromSession", "string", ["string"]),
        destroySession: library.func("destroySession", "string", ["string"]),
        destroyAll: library.func("destroyAll", "string", []),
        freeMemory: library.func("freeMemory", "void", ["string"]),
    };
    const binding = {
        libraryPath,
        async request(payload) {
            return parseNativeResponse(await invokeNativeAsync("request", rawBinding.request, JSON.stringify(payload)), "request", rawBinding);
        },
        async getCookiesFromSession(payload) {
            return parseNativeResponse(await invokeNativeAsync("getCookiesFromSession", rawBinding.getCookiesFromSession, JSON.stringify(payload)), "getCookiesFromSession", rawBinding);
        },
        async destroySession(payload) {
            return parseNativeResponse(await invokeNativeAsync("destroySession", rawBinding.destroySession, JSON.stringify(payload)), "destroySession", rawBinding);
        },
        async destroyAll() {
            return parseNativeResponse(await invokeNativeAsync("destroyAll", rawBinding.destroyAll), "destroyAll", rawBinding);
        },
    };
    bindingCache.set(libraryPath, binding);
    return binding;
}
async function ensureNativeBinding(options = {}) {
    const library = await (0, binary_1.ensureNativeLibrary)(options);
    return loadNativeBinding(library.libraryPath);
}
