// @lobehub/ui/es/Markdown/SyntaxMarkdown/StreamdownRender.mjs
"use client";
import { isDeepEqual } from "../../utils/isDeepEqual.mjs";
import { useMarkdownContext } from "../components/MarkdownProvider.mjs";
import { useMarkdownComponents } from "../../hooks/useMarkdown/useMarkdownComponents.mjs";
import { useMarkdownContent } from "../../hooks/useMarkdown/useMarkdownContent.mjs";
import { useMarkdownRehypePlugins } from "../../hooks/useMarkdown/useMarkdownRehypePlugins.mjs";
import { useMarkdownRemarkPlugins } from "../../hooks/useMarkdown/useMarkdownRemarkPlugins.mjs";
import { rehypeStreamAnimated } from "../plugins/rehypeStreamAnimated.mjs";
import { useStreamdownProfiler } from "../streamProfiler/StreamdownProfilerProvider.mjs";
import { resolveBlockAnimationMeta } from "./streamAnimationMeta.mjs";
import { styles } from "./style.mjs";
import { useSmoothStreamContent } from "./useSmoothStreamContent.mjs";
import { useStreamQueue } from "./useStreamQueue.mjs";
import { Profiler, createElement, memo, useCallback, useEffect, useId, useMemo, useRef } from "react";
import { jsx } from "react/jsx-runtime";
import Markdown from "react-markdown";
import { marked } from "marked";
import remend from "remend";
//#region src/Markdown/SyntaxMarkdown/StreamdownRender.tsx
const STREAM_FADE_DURATION = 280;
const REVEALED_STREAM_PLUGIN = [rehypeStreamAnimated, { revealed: true }];
function countChars(text) {
	return [...text].length;
}
function getNow() {
	return typeof performance === "undefined" ? Date.now() : performance.now();
}
const isSamePlugin = (prevPlugin, nextPlugin) => {
	const prevTuple = Array.isArray(prevPlugin) ? prevPlugin : [prevPlugin];
	const nextTuple = Array.isArray(nextPlugin) ? nextPlugin : [nextPlugin];
	if (prevTuple.length !== nextTuple.length) return false;
	if (prevTuple[0] !== nextTuple[0]) return false;
	return isDeepEqual(prevTuple.slice(1), nextTuple.slice(1));
};
const isSamePlugins = (prevPlugins, nextPlugins) => {
	if (prevPlugins === nextPlugins) return true;
	if (!prevPlugins || !nextPlugins) return !prevPlugins && !nextPlugins;
	if (prevPlugins.length !== nextPlugins.length) return false;
	for (let i = 0; i < prevPlugins.length; i++) if (!isSamePlugin(prevPlugins[i], nextPlugins[i])) return false;
	return true;
};
const useStablePlugins = (plugins) => {
	const stableRef = useRef(plugins);
	if (!isSamePlugins(stableRef.current, plugins)) stableRef.current = plugins;
	return stableRef.current;
};
const StreamdownBlock = memo(({ children, ...rest }) => {
	return /* @__PURE__ */ jsx(Markdown, {
		...rest,
		children
	});
}, (prevProps, nextProps) => prevProps.children === nextProps.children && prevProps.components === nextProps.components && isSamePlugins(prevProps.rehypePlugins, nextProps.rehypePlugins) && isSamePlugins(prevProps.remarkPlugins, nextProps.remarkPlugins));
StreamdownBlock.displayName = "StreamdownBlock";
const StreamdownRender = memo(({ children, ...rest }) => {
	const { streamSmoothingPreset = "balanced" } = useMarkdownContext();
	const profiler = useStreamdownProfiler();
	const escapedContent = useMarkdownContent(children || "");
	const components = useMarkdownComponents();
	const baseRehypePlugins = useStablePlugins(useMarkdownRehypePlugins());
	const remarkPlugins = useStablePlugins(useMarkdownRemarkPlugins());
	const generatedId = useId();
	const smoothedContent = useSmoothStreamContent(typeof escapedContent === "string" ? escapedContent : "", { preset: streamSmoothingPreset });
	const processedContentResult = useMemo(() => {
		const start = profiler ? getNow() : 0;
		const value = remend(smoothedContent);
		return {
			durationMs: profiler ? getNow() - start : 0,
			value
		};
	}, [profiler, smoothedContent]);
	const processedContent = processedContentResult.value;
	const blocksResult = useMemo(() => {
		const start = profiler ? getNow() : 0;
		const tokens = marked.lexer(processedContent);
		let offset = 0;
		const value = tokens.map((token) => {
			const block = {
				content: token.raw,
				startOffset: offset
			};
			offset += token.raw.length;
			return block;
		});
		return {
			durationMs: profiler ? getNow() - start : 0,
			value
		};
	}, [processedContent, profiler]);
	const blocks = blocksResult.value;
	const { getBlockState, charDelay } = useStreamQueue(blocks);
	const blockCharDelayRef = useRef(/* @__PURE__ */ new Map());
	const blockBirthsRef = useRef(/* @__PURE__ */ new Map());
	const renderNow = getNow();
	const birthsResult = useMemo(() => {
		const start = profiler ? getNow() : 0;
		const nextBirths = /* @__PURE__ */ new Map();
		const prevBirths = blockBirthsRef.current;
		for (const [index, block] of blocks.entries()) {
			if (getBlockState(index) === "queued") continue;
			const blockCharCount = countChars(block.content);
			const prev = prevBirths.get(block.startOffset);
			let arr;
			if (prev && prev.length === blockCharCount) arr = prev;
			else if (prev && prev.length > blockCharCount) arr = prev.slice(0, blockCharCount);
			else {
				arr = prev ? prev.slice() : [];
				const startIdx = arr.length;
				const cap = renderNow + STREAM_FADE_DURATION;
				for (let i = startIdx; i < blockCharCount; i++) {
					const chained = (i > 0 ? arr[i - 1] : renderNow - charDelay) + charDelay;
					arr.push(Math.min(cap, Math.max(chained, renderNow)));
				}
			}
			nextBirths.set(block.startOffset, arr);
		}
		return {
			durationMs: profiler ? getNow() - start : 0,
			value: nextBirths
		};
	}, [
		blocks,
		charDelay,
		getBlockState,
		profiler,
		renderNow
	]);
	const birthsForRender = birthsResult.value;
	useEffect(() => {
		if (!profiler) return;
		profiler.recordCalculation({
			durationMs: processedContentResult.durationMs,
			name: "content-normalize",
			textLength: processedContent.length
		});
	}, [
		processedContent.length,
		processedContentResult.durationMs,
		profiler
	]);
	useEffect(() => {
		if (!profiler) return;
		profiler.recordCalculation({
			durationMs: blocksResult.durationMs,
			itemCount: blocks.length,
			name: "block-lex",
			textLength: processedContent.length
		});
	}, [
		blocks.length,
		blocksResult.durationMs,
		processedContent.length,
		profiler
	]);
	useEffect(() => {
		if (!profiler) return;
		profiler.recordCalculation({
			durationMs: birthsResult.durationMs,
			itemCount: blocks.length,
			name: "block-births",
			textLength: processedContent.length
		});
	}, [
		birthsResult.durationMs,
		blocks.length,
		processedContent.length,
		profiler
	]);
	const blockAnimationMetaResult = useMemo(() => {
		const nextBlockCharDelay = /* @__PURE__ */ new Map();
		const blockAnimationMeta = /* @__PURE__ */ new Map();
		for (const [index, block] of blocks.entries()) {
			const state = getBlockState(index);
			const births = birthsForRender.get(block.startOffset);
			const animationMeta = resolveBlockAnimationMeta({
				currentCharDelay: charDelay,
				fadeDuration: STREAM_FADE_DURATION,
				lastElapsedMs: renderNow - (births && births.length > 0 ? births.at(-1) ?? renderNow : renderNow),
				previousCharDelay: blockCharDelayRef.current.get(block.startOffset),
				state
			});
			nextBlockCharDelay.set(block.startOffset, animationMeta.charDelay);
			blockAnimationMeta.set(block.startOffset, animationMeta);
		}
		return {
			blockAnimationMeta,
			blockCharDelay: nextBlockCharDelay
		};
	}, [
		birthsForRender,
		blocks,
		charDelay,
		getBlockState,
		renderNow
	]);
	useEffect(() => {
		blockCharDelayRef.current = blockAnimationMetaResult.blockCharDelay;
		blockBirthsRef.current = birthsForRender;
	}, [birthsForRender, blockAnimationMetaResult.blockCharDelay]);
	const handleRootRender = useCallback((_, phase, actualDuration, baseDuration) => {
		profiler?.recordRootCommit({
			actualDuration,
			baseDuration,
			blockCount: blocks.length,
			phase,
			textLength: processedContent.length
		});
	}, [
		blocks.length,
		processedContent.length,
		profiler
	]);
	const handleBlockRender = useCallback((id, phase, actualDuration, baseDuration) => {
		if (!profiler) return;
		const [, indexText, offsetText] = id.split(":");
		const blockIndex = Number(indexText);
		if (!Number.isFinite(blockIndex)) return;
		const block = blocks[blockIndex];
		if (!block) return;
		profiler.recordBlockCommit({
			actualDuration,
			baseDuration,
			blockChars: countChars(block.content),
			blockIndex,
			blockKey: offsetText ?? String(block.startOffset),
			phase,
			state: getBlockState(blockIndex)
		});
	}, [
		blocks,
		getBlockState,
		profiler
	]);
	const content = /* @__PURE__ */ jsx("div", {
		className: styles.animated,
		children: blocks.map((block, index) => {
			if (getBlockState(index) === "queued") return null;
			const animationMeta = blockAnimationMetaResult.blockAnimationMeta.get(block.startOffset);
			if (!animationMeta) return null;
			const births = birthsForRender.get(block.startOffset);
			const plugins = animationMeta.settled ? [...baseRehypePlugins, REVEALED_STREAM_PLUGIN] : [...baseRehypePlugins, [rehypeStreamAnimated, {
				births,
				fadeDuration: STREAM_FADE_DURATION,
				nowMs: renderNow
			}]];
			const key = `${generatedId}-${block.startOffset}`;
			const blockNode = /* @__PURE__ */ jsx(StreamdownBlock, {
				...rest,
				components,
				rehypePlugins: plugins,
				remarkPlugins,
				children: block.content
			});
			if (!profiler) return /* @__PURE__ */ createElement(StreamdownBlock, {
				...rest,
				components,
				key,
				rehypePlugins: plugins,
				remarkPlugins
			}, block.content);
			return /* @__PURE__ */ jsx(Profiler, {
				id: `streamdown-block:${index}:${block.startOffset}`,
				onRender: handleBlockRender,
				children: blockNode
			}, key);
		})
	});
	if (!profiler) return content;
	return /* @__PURE__ */ jsx(Profiler, {
		id: "streamdown-root",
		onRender: handleRootRender,
		children: content
	});
});
StreamdownRender.displayName = "StreamdownRender";
//#endregion
export { StreamdownRender as default };
