// @lobehub/ui/es/index.mjs
import { MotionComponent, MotionProvider, useMotionComponent } from "./MotionProvider/index.mjs";
import { genCdnUrl } from "./utils/genCdnUrl.mjs";
import ConfigProvider, { useCdnFn } from "./ConfigProvider/index.mjs";
import A from "./A/index.mjs";
import Accordion from "./Accordion/Accordion.mjs";
import FlexBasic_default from "./Flex/FlexBasic.mjs";
import Center from "./Flex/Center.mjs";
import { findCustomThemeName, neutralColors, neutralColorsSwatches, primaryColors, primaryColorsSwatches } from "./styles/customTheme.mjs";
import { generateCustomStylish } from "./styles/theme/customStylish.mjs";
import { staticStylish } from "./styles/theme/customStylishStatic.mjs";
import { generateCustomToken } from "./styles/theme/customToken.mjs";
import { generateColorNeutralPalette, generateColorPalette } from "./styles/theme/generateColorPalette.mjs";
import Block from "./Block/Block.mjs";
import { useAppElement } from "./ThemeProvider/AppElementContext.mjs";
import { LOBE_THEME_APP_ID } from "./ThemeProvider/constants.mjs";
import Meta from "./ThemeProvider/Meta.mjs";
import FontLoader from "./FontLoader/index.mjs";
import { CLASSNAMES } from "./styles/classNames.mjs";
import ThemeProvider from "./ThemeProvider/ThemeProvider.mjs";
import { placementMap, toFloatingUIPlacement } from "./utils/placement.mjs";
import { KeyMapEnum } from "./Hotkey/const.mjs";
import { IconProvider } from "./Icon/components/IconProvider.mjs";
import Icon from "./Icon/Icon.mjs";
import { combineKeys } from "./Hotkey/utils.mjs";
import Hotkey from "./Hotkey/Hotkey.mjs";
import Tooltip from "./Tooltip/Tooltip.mjs";
import TooltipGroup from "./Tooltip/TooltipGroup.mjs";
import Text from "./Text/Text.mjs";
import { preventDefault, preventDefaultAndStopPropagation, stopPropagation } from "./utils/dom.mjs";
import AccordionItem from "./Accordion/AccordionItem.mjs";
import ActionIcon from "./ActionIcon/ActionIcon.mjs";
import { styles } from "./Menu/sharedStyle.mjs";
import { DropdownMenuCheckboxItemIndicator, DropdownMenuCheckboxItemPrimitive, DropdownMenuGroup, DropdownMenuGroupLabel, DropdownMenuItem, DropdownMenuItemContent, DropdownMenuItemExtra, DropdownMenuItemIcon, DropdownMenuItemLabel, DropdownMenuPopup, DropdownMenuPortal, DropdownMenuPositioner, DropdownMenuRoot, DropdownMenuSeparator, DropdownMenuSubmenuArrow, DropdownMenuSubmenuRoot, DropdownMenuSubmenuTrigger, DropdownMenuTrigger } from "./base-ui/DropdownMenu/atoms.mjs";
import Menu from "./Menu/Menu.mjs";
import { renderDropdownMenuItems } from "./base-ui/DropdownMenu/renderItems.mjs";
import DropdownMenu from "./base-ui/DropdownMenu/DropdownMenu.mjs";
import ActionIconGroup from "./ActionIconGroup/ActionIconGroup.mjs";
import Alert from "./Alert/Alert.mjs";
import AutoComplete from "./AutoComplete/Select.mjs";
import FluentEmoji from "./FluentEmoji/FluentEmoji.mjs";
import AvatarGroup from "./Avatar/AvatarGroup/index.mjs";
import Avatar from "./Avatar/index.mjs";
import Burger from "./Burger/Burger.mjs";
import Button from "./Button/Button.mjs";
import CheckboxGroup from "./Checkbox/CheckboxGroup.mjs";
import Checkbox from "./Checkbox/index.mjs";
import FileTypeIcon from "./FileTypeIcon/FileTypeIcon.mjs";
import MaterialFileTypeIcon from "./MaterialFileTypeIcon/MaterialFileTypeIcon.mjs";
import CodeDiff from "./CodeDiff/CodeDiff.mjs";
import PatchDiff from "./CodeDiff/PatchDiff.mjs";
import { highlighterThemes } from "./Highlighter/const.mjs";
import lobe_theme_default from "./Highlighter/theme/lobe-theme.mjs";
import SyntaxHighlighter from "./Highlighter/SyntaxHighlighter/index.mjs";
import CodeEditor from "./CodeEditor/CodeEditor.mjs";
import Collapse from "./Collapse/Collapse.mjs";
import ColorSwatches from "./ColorSwatches/ColorSwatches.mjs";
import { I18nProvider } from "./i18n/context.mjs";
import { useTranslation } from "./i18n/useTranslation.mjs";
import { closeContextMenu, showContextMenu, updateContextMenuItems } from "./base-ui/ContextMenu/store.mjs";
import { ContextMenuHost } from "./base-ui/ContextMenu/ContextMenuHost.mjs";
import { ContextMenuTrigger } from "./base-ui/ContextMenu/ContextMenuTrigger.mjs";
import { copyToClipboard } from "./utils/copyToClipboard.mjs";
import CopyButton from "./CopyButton/CopyButton.mjs";
import DatePicker from "./DatePicker/DatePicker.mjs";
import DownloadButton from "./DownloadButton/DownloadButton.mjs";
import DraggablePanelBody from "./DraggablePanel/components/DraggablePanelBody.mjs";
import DraggablePanelContainer from "./DraggablePanel/components/DraggablePanelContainer.mjs";
import DraggablePanelFooter from "./DraggablePanel/components/DraggablePanelFooter.mjs";
import DraggablePanelHeader from "./DraggablePanel/components/DraggablePanelHeader.mjs";
import DraggablePanel from "./DraggablePanel/index.mjs";
import DraggableSideNav from "./DraggableSideNav/DraggableSideNav.mjs";
import Drawer from "./Drawer/Drawer.mjs";
import Dropdown from "./Dropdown/Dropdown.mjs";
import Input from "./Input/Input.mjs";
import InputNumber from "./Input/InputNumber.mjs";
import InputOPT from "./Input/InputOPT.mjs";
import InputPassword from "./Input/InputPassword.mjs";
import TextArea from "./Input/TextArea.mjs";
import EditableText from "./EditableText/EditableText.mjs";
import EditorSlashMenu from "./EditorSlashMenu/EditorSlashMenu.mjs";
import { PopoverArrowIcon } from "./base-ui/Popover/ArrowIcon.mjs";
import { usePopoverPortalContainer } from "./base-ui/Popover/PopoverPortal.mjs";
import { PopoverArrow, PopoverBackdrop, PopoverPopup, PopoverPortal, PopoverPositioner, PopoverRoot, PopoverTriggerElement, PopoverViewport } from "./base-ui/Popover/atoms.mjs";
import { PopoverProvider, usePopoverContext } from "./base-ui/Popover/context.mjs";
import Popover from "./base-ui/Popover/Popover.mjs";
import PopoverGroup from "./base-ui/Popover/PopoverGroup.mjs";
import Tabs from "./Tabs/Tabs.mjs";
import Tag from "./Tag/Tag.mjs";
import EmojiPicker from "./EmojiPicker/EmojiPicker.mjs";
import Empty from "./Empty/Empty.mjs";
import Footer from "./Footer/Footer.mjs";
import FormGroup from "./Form/components/FormGroup.mjs";
import FormTitle from "./Form/components/FormTitle.mjs";
import FormItem from "./Form/components/FormItem.mjs";
import FormSubmitFooter from "./Form/components/FormSubmitFooter.mjs";
import Form from "./Form/index.mjs";
import Modal from "./Modal/Modal.mjs";
import { ModalProvider, useModalContext } from "./Modal/ModalProvider.mjs";
import { ModalHost, createModal, createRawModal } from "./Modal/imperative.mjs";
import FormModal from "./FormModal/FormModal.mjs";
import Freeze from "./Freeze/Freeze.mjs";
import Grid from "./Grid/Grid.mjs";
import GroupAvatar from "./GroupAvatar/GroupAvatar.mjs";
import GuideCard from "./GuideCard/GuideCard.mjs";
import Header from "./Header/Header.mjs";
import Highlighter from "./Highlighter/Highlighter.mjs";
import { preprocessMarkdownContent } from "./hooks/useMarkdown/utils.mjs";
import HotkeyInput from "./HotkeyInput/HotkeyInput.mjs";
import { DEFAULT_HEIGHT, DEFAULT_SANDBOX, containsScript, isFullHtmlDocument, isHtmlContentClosed } from "./HtmlPreview/const.mjs";
import NeuralNetworkLoading from "./NeuralNetworkLoading/NeuralNetworkLoading.mjs";
import Segmented from "./Segmented/Segmented.mjs";
import { AUTO_HEIGHT_MESSAGE_TYPE } from "./HtmlPreview/injectAutoHeightScript.mjs";
import HtmlPreviewIframe from "./HtmlPreview/Iframe.mjs";
import HtmlPreview from "./HtmlPreview/HtmlPreview.mjs";
import PreviewGroup from "./Image/PreviewGroup.mjs";
import Image from "./Image/index.mjs";
import ImageSelect from "./ImageSelect/ImageSelect.mjs";
import LayoutFooter from "./Layout/components/LayoutFooter.mjs";
import LayoutHeader from "./Layout/components/LayoutHeader.mjs";
import LayoutMain from "./Layout/components/LayoutMain.mjs";
import LayoutSidebar from "./Layout/components/LayoutSidebar.mjs";
import LayoutSidebarInner from "./Layout/components/LayoutSidebarInner.mjs";
import LayoutToc from "./Layout/components/LayoutToc.mjs";
import Layout from "./Layout/index.mjs";
import ListItem from "./List/ListItem/index.mjs";
import List from "./List/index.mjs";
import { mermaidThemes } from "./Mermaid/const.mjs";
import SyntaxMermaid from "./Mermaid/SyntaxMermaid/index.mjs";
import Mermaid from "./Mermaid/Mermaid.mjs";
import Snippet from "./Snippet/Snippet.mjs";
import ScrollShadow from "./ScrollShadow/ScrollShadow.mjs";
import SearchResultCards from "./Markdown/components/SearchResultCards/index.mjs";
import Video from "./Video/index.mjs";
import { rehypeCustomFootnotes } from "./Markdown/plugins/rehypeCustomFootnotes.mjs";
import { rehypeKatexDir } from "./Markdown/plugins/rehypeKatexDir.mjs";
import { remarkBr } from "./Markdown/plugins/remarkBr.mjs";
import { remarkCustomFootnotes } from "./Markdown/plugins/remarkCustomFootnotes.mjs";
import { remarkGfmPlus } from "./Markdown/plugins/remarkGfmPlus.mjs";
import { remarkVideo } from "./Markdown/plugins/remarkVideo.mjs";
import { rehypeStreamAnimated } from "./Markdown/plugins/rehypeStreamAnimated.mjs";
import Typography from "./Markdown/Typography.mjs";
import Markdown from "./Markdown/Markdown.mjs";
import { remarkColor } from "./Markdown/plugins/remarkColor.mjs";
import MaskShadow from "./MaskShadow/MaskShadow.mjs";
import { ScrollAreaContent, ScrollAreaCorner, ScrollAreaRoot, ScrollAreaScrollbar, ScrollAreaThumb, ScrollAreaViewport } from "./base-ui/ScrollArea/atoms.mjs";
import { ScrollArea } from "./base-ui/ScrollArea/ScrollArea.mjs";
import SearchBar from "./SearchBar/SearchBar.mjs";
import Select from "./Select/Select.mjs";
import SideNav from "./SideNav/SideNav.mjs";
import SkeletonBlock from "./Skeleton/SkeletonBlock.mjs";
import SkeletonAvatar from "./Skeleton/SkeletonAvatar.mjs";
import SkeletonParagraph from "./Skeleton/SkeletonParagraph.mjs";
import SkeletonTitle from "./Skeleton/SkeletonTitle.mjs";
import SkeletonButton from "./Skeleton/SkeletonButton.mjs";
import SkeletonTags from "./Skeleton/SkeletonTags.mjs";
import Skeleton from "./Skeleton/index.mjs";
import SliderWithInput from "./SliderWithInput/SliderWithInput.mjs";
import SortableList from "./SortableList/SortableList.mjs";
import ThemeSwitch from "./ThemeSwitch/ThemeSwitch.mjs";
import { ToastHost, toast, useToast } from "./base-ui/Toast/imperative.mjs";
import Toc from "./Toc/Toc.mjs";
import { ErrorBoundary } from "react-error-boundary";
export { A, Accordion, AccordionItem, ActionIcon, ActionIconGroup, Alert, AutoComplete, Avatar, AvatarGroup, Block, Burger, Button, CLASSNAMES, Center, Checkbox, CheckboxGroup, CodeDiff, CodeEditor, Collapse, ColorSwatches, ConfigProvider, ContextMenuHost, ContextMenuTrigger, CopyButton, DatePicker, DownloadButton, DraggablePanel, DraggablePanelBody, DraggablePanelContainer, DraggablePanelFooter, DraggablePanelHeader, DraggableSideNav, Drawer, Dropdown, DropdownMenu, DropdownMenuCheckboxItemIndicator, DropdownMenuCheckboxItemPrimitive, DropdownMenuGroup, DropdownMenuGroupLabel, DropdownMenuItem, DropdownMenuItemContent, DropdownMenuItemExtra, DropdownMenuItemIcon, DropdownMenuItemLabel, DropdownMenuPopup, DropdownMenuPortal, DropdownMenuPositioner, DropdownMenuRoot, DropdownMenuSeparator, DropdownMenuSubmenuArrow, DropdownMenuSubmenuRoot, DropdownMenuSubmenuTrigger, DropdownMenuTrigger, EditableText, EditorSlashMenu, EmojiPicker, Empty, ErrorBoundary, FileTypeIcon, FlexBasic_default as FlexBasic, FlexBasic_default as Flexbox, FluentEmoji, FontLoader, Footer, Form, FormGroup, FormItem, FormModal, FormSubmitFooter, FormTitle, Freeze, Grid, GroupAvatar, GuideCard, DEFAULT_HEIGHT as HTML_PREVIEW_DEFAULT_HEIGHT, DEFAULT_SANDBOX as HTML_PREVIEW_DEFAULT_SANDBOX, AUTO_HEIGHT_MESSAGE_TYPE as HTML_PREVIEW_RESIZE_MESSAGE, Header, Highlighter, Hotkey, HotkeyInput, HtmlPreview, HtmlPreviewIframe, I18nProvider, Icon, IconProvider, Image, ImageSelect, Input, InputNumber, InputOPT, InputPassword, KeyMapEnum, LOBE_THEME_APP_ID, Layout, LayoutFooter, LayoutHeader, LayoutMain, LayoutSidebar, LayoutSidebarInner, LayoutToc, List, ListItem, I18nProvider as LobeUIProvider, Markdown, MaskShadow, MaterialFileTypeIcon, Menu, Mermaid, Meta, Modal, ModalHost, ModalProvider, MotionComponent, MotionProvider, NeuralNetworkLoading, PatchDiff, Popover, PopoverArrow, PopoverArrowIcon, PopoverBackdrop, PopoverGroup, PopoverPopup, PopoverPortal, PopoverPositioner, PopoverProvider, PopoverRoot, PopoverTriggerElement, PopoverViewport, PreviewGroup, ScrollArea, ScrollAreaContent, ScrollAreaCorner, ScrollAreaRoot, ScrollAreaScrollbar, ScrollAreaThumb, ScrollAreaViewport, ScrollShadow, SearchBar, SearchResultCards, Segmented, Select, lobe_theme_default as ShikiLobeTheme, SideNav, Skeleton, SkeletonAvatar, SkeletonBlock, SkeletonButton, SkeletonParagraph, SkeletonTags, SkeletonTitle, SliderWithInput, Snippet, SortableList, SyntaxHighlighter, SyntaxMermaid, Tabs, Tag, Text, TextArea, ThemeProvider, ThemeSwitch, ToastHost, Toc, Tooltip, TooltipGroup, Typography, Video, closeContextMenu, combineKeys, copyToClipboard, createModal, createRawModal, findCustomThemeName, genCdnUrl, generateColorNeutralPalette, generateColorPalette, highlighterThemes, containsScript as htmlPreviewContainsScript, isFullHtmlDocument, isHtmlContentClosed, generateCustomStylish as lobeCustomStylish, generateCustomToken as lobeCustomToken, staticStylish as lobeStaticStylish, styles as menuSharedStyles, mermaidThemes, neutralColors, neutralColorsSwatches, placementMap, preprocessMarkdownContent, preventDefault, preventDefaultAndStopPropagation, primaryColors, primaryColorsSwatches, rehypeCustomFootnotes, rehypeKatexDir, rehypeStreamAnimated, remarkBr, remarkColor, remarkCustomFootnotes, remarkGfmPlus, remarkVideo, renderDropdownMenuItems, showContextMenu, stopPropagation, toFloatingUIPlacement, toast, updateContextMenuItems, useAppElement, useCdnFn, useModalContext, useMotionComponent, usePopoverContext, usePopoverPortalContainer, useToast, useTranslation };
