// tls-client-node/dist/multipart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultipartForm = void 0;
exports.createMultipartForm = createMultipartForm;
function isMultipartFileValue(value) {
    return typeof value === "object" && value !== null && "data" in value;
}
function createBlobFromSource(value, contentType) {
    if (value instanceof Blob && contentType === undefined) {
        return value;
    }
    if (value instanceof Blob) {
        return new Blob([value], { type: contentType });
    }
    if (value instanceof ArrayBuffer) {
        return new Blob([value], { type: contentType });
    }
    if (ArrayBuffer.isView(value)) {
        const bytes = Uint8Array.from(new Uint8Array(value.buffer, value.byteOffset, value.byteLength));
        return new Blob([
            bytes,
        ], { type: contentType });
    }
    return new Blob([value], { type: contentType });
}
class MultipartForm {
    formData = new FormData();
    append(name, value) {
        if (value instanceof Blob) {
            this.formData.append(name, value);
        }
        else {
            this.formData.append(name, String(value));
        }
        return this;
    }
    appendFile(name, value, options) {
        this.formData.append(name, createBlobFromSource(value, options.contentType), options.filename);
        return this;
    }
    appendJson(name, value, filename = `${name}.json`) {
        return this.appendFile(name, JSON.stringify(value), {
            filename,
            contentType: "application/json",
        });
    }
    toFormData() {
        return this.formData;
    }
    static fromRecord(record) {
        const form = new MultipartForm();
        for (const [name, rawValue] of Object.entries(record)) {
            const values = Array.isArray(rawValue) ? rawValue : [rawValue];
            for (const value of values) {
                if (isMultipartFileValue(value)) {
                    form.appendFile(name, value.data, value);
                }
                else {
                    form.append(name, value);
                }
            }
        }
        return form;
    }
}
exports.MultipartForm = MultipartForm;
function createMultipartForm(record) {
    return record ? MultipartForm.fromRecord(record) : new MultipartForm();
}
