// tls-client-node/dist/response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TLSResponse = void 0;
function isDataUrl(value) {
    // Strict full-string match for the exact format the Go library emits for byte
    // responses. The loose startsWith+includes check would false-positive on text
    // bodies that happen to contain "data:" and ";base64," (e.g. CSS with data URIs).
    return /^data:[^;]{1,256};base64,[A-Za-z0-9+/]*={0,2}$/.test(value);
}
class TLSResponse {
    id;
    ok;
    status;
    url;
    body;
    headers;
    cookies;
    sessionId;
    usedProtocol;
    constructor(payload) {
        this.id = payload.id;
        this.ok = payload.status >= 200 && payload.status < 300;
        this.status = payload.status;
        this.url = payload.target;
        this.body = payload.body;
        // Normalize to lowercase so header() lookups are case-insensitive regardless
        // of what casing the Go API uses in its response.
        this.headers = Object.fromEntries(Object.entries(payload.headers ?? {}).map(([k, v]) => [k.toLowerCase(), v]));
        this.cookies = payload.cookies ?? {};
        this.sessionId = payload.sessionId;
        this.usedProtocol = payload.usedProtocol;
    }
    text() {
        return Promise.resolve(this.body);
    }
    json() {
        return Promise.resolve(JSON.parse(this.body));
    }
    bytes() {
        if (!isDataUrl(this.body)) {
            return Promise.resolve(new TextEncoder().encode(this.body));
        }
        const [, encoded = ""] = this.body.split(",", 2);
        return Promise.resolve(Buffer.from(encoded, "base64"));
    }
    async arrayBuffer() {
        const bytes = await this.bytes();
        return Uint8Array.from(bytes).buffer;
    }
    header(name) {
        return this.headers[name.toLowerCase()]?.join(", ");
    }
    toJSON() {
        return {
            id: this.id,
            status: this.status,
            target: this.url,
            body: this.body,
            headers: this.headers,
            cookies: this.cookies,
            sessionId: this.sessionId,
            usedProtocol: this.usedProtocol,
        };
    }
}
exports.TLSResponse = TLSResponse;
